File: src/main/java/se/lth/base/server/data/Drive.java
package se.lth.base.server.data;

/**
 * Data class for a drive
 *
 * @author Group 1 ETSN05 2018
 * @see DriveDataAccess
 */
public class Drive {
    private final int driveId;
    private final int carNumberOfSeats;
    private final int optLuggageSize;
    private final String start;
    private final String stop;
    private final String comment;
    private final String carBrand;
    private final String carModel;
    private final String carColor;
    private final String carLicensePlate;
    private final boolean optWinterTires;
    private final boolean optPets;
    private final boolean optBicycle;
    private final long departureTime;
    private final long arrivalTime;

    /**
     * @param driveId          the Id of the drive.
     * @param start            the start location of the drive.
     * @param stop             the final destination of the drive.
     * @param departureTime    the time and date of the departure.
     * @param arrivalTime      the time and date of arrival
     * @param comment          additional comment, could be used when extra information about a drive is needed.
     * @param carBrand         the brand of the car.
     * @param carModel         the model of the car.
     * @param carColor         the color of the car.
     * @param carLicensePlate  the license plate of the car.
     * @param carNumberOfSeats the number of seats in the car.
     * @param optLuggageSize   the size of the luggage in the car. Can take values 0,1 or 2 (referring to small, medium and large).
     * @param optWinterTires   specifies if the car has winter tiers or not.
     * @param optBicycle       specifies if it is possible to bring a bicycle.
     * @param optPets          specifies if pets are allowed in the car.
     */
    public Drive(int driveId, String start, String stop, long departureTime, long arrivalTime, String comment, String carBrand,
                 String carModel, String carColor, String carLicensePlate, int carNumberOfSeats, int optLuggageSize,
                 boolean optWinterTires, boolean optBicycle, boolean optPets) {
        this.driveId = driveId;
        this.start = start;
        this.stop = stop;
        this.departureTime = departureTime;
        this.arrivalTime = arrivalTime;
        this.comment = comment;
        this.carBrand = carBrand;
        this.carModel = carModel;
        this.carColor = carColor;
        this.carLicensePlate = carLicensePlate;
        this.carNumberOfSeats = carNumberOfSeats;
        this.optLuggageSize = optLuggageSize;
        this.optWinterTires = optWinterTires;
        this.optBicycle = optBicycle;
        this.optPets = optPets;
    }

    /**
     * Returns the driveId.
     */
    public int getDriveId() {
        return driveId;
    }

    /**
     * Returns the start location.
     */
    public String getStart() {
        return start;
    }

    /**
     * Returns the stop location.
     */
    public String getStop() {
        return stop;
    }

    /**
     * Returns the departure time.
     */
    public long getDepartureTime() {
        return departureTime;
    }

    /**
     * Returns the arrival time.
     */
    public long getArrivalTime() {
        return arrivalTime;
    }

    /**
     * Returns the comment of a drive.
     */
    public String getComment() {
        return comment;
    }

    /**
     * Returns the brand of the car.
     */
    String getCarBrand() {
        return carBrand;
    }

    /**
     * Returns the model of the car.
     */
    String getCarModel() {
        return carModel;
    }

    /**
     * Returns the color of the car.
     */
    public String getCarColor() {
        return carColor;
    }

    /**
     * Returns the license plate of the car.
     */
    public String getCarLicensePlate() {
        return carLicensePlate;
    }

    /**
     * Returns the number of seats in the car.
     */
    public int getCarNumberOfSeats() {
        return carNumberOfSeats;
    }

    /**
     * Returns the size of the luggage in the car.
     */
    int getOptLuggageSize() {
        return optLuggageSize;
    }

    /**
     * Returns true if the car has winter tires, otherwise false.
     */
    boolean getOptWinterTires() {
        return optWinterTires;
    }

    /**
     * Returns true if it is possible to bring a bicycle, otherwise false.
     */
    boolean getOptBicycle() {
        return optBicycle;
    }

    /**
     * Returns true if pets are allowed in the car, otherwise false.
     */
    boolean getOptPets() {
        return optPets;
    }
}

File: src/main/resources/webassets/js/custom/templates/drive.js
window.base = window.base || {};

window.base.driveController = (() => {
    let model = {
        user: {},
        driveUser: {},
        driveWrap: {},
        driveReport: {}
    };

    const view = {
        showReportModal: () => {
            new Modal(document.getElementById('report-modal'), {keyboard: false}).show();
        },

        showFailure: (msg, element) => {
            let alertClasses = element.classList;

            alertClasses.remove('d-none');
            alertClasses.remove('alert-info');
            alertClasses.add('alert-danger');

            element.innerHTML = `<h5 class="alert-heading">Oops!</h5><p>Something went wrong, error message: ${msg}.</p>`;
        },

        showSuccess: (msg, element) => {
            let alertClasses = element.classList;

            alertClasses.remove('d-none');
            alertClasses.remove('alert-danger');
            alertClasses.add('alert-info');

            element.innerHTML = `<h5 class="alert-heading">Done</h5><p>${msg}</p>`;
        },

        renderDrive: () => {
            const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
            const drive = model.driveWrap.drive;
            const driveUsers = model.driveWrap.users;
            let driver;
            let passengerHtml = '';
            let acceptedPassengers = 1;

            for (let i = 0; i < driveUsers.length; i++) {
                if (driveUsers[i].driver) {
                    driver = driveUsers[i].info;
                }
            }

            const today = new Date();
            const dob = new Date(driver.dateOfBirth);
            let driverAge = today.getFullYear() - dob.getFullYear();
            const m = today.getMonth() - dob.getMonth();
            if (m < 0 || (m === 0 && today.getDate() < dob.getDate())) {
                driverAge--;
            }

            let noDrives;
            let driverReviews;
            let driverRating;

            const requestButton = document.getElementById('drive-request');

            for (let i = 0; i < driveUsers.length; i++) {
                const user = driveUsers[i];
                const firstName = user.info.firstName;
                const score = user.info.ratingTotalScore;
                const reviews = user.info.numberOfRatings;
                const rating = reviews === 0 ? 'No' : parseFloat(score / reviews).toFixed(2);

                if (user.userId === model.user.userId) {
                    requestButton.setAttribute('disabled', '');
                    requestButton.textContent = 'Requested';
                }

                if (user.userId === model.user.userId && user.driver) {
                    requestButton.textContent = 'Driving';
                }

                if (user.driver) {
                    noDrives = user.noDrives;
                    driverReviews = reviews;
                    driverRating = rating;
                } else if (user.accepted) {
                    acceptedPassengers++;
                    passengerHtml += `${firstName} (<i class="fas fa-star fa-sm">${rating}), `;
                }
            }

            passengerHtml = passengerHtml.length === 0 ? 'None' : passengerHtml.slice(0, passengerHtml.length - 2);

            const dtd = new Date(drive.departureTime);
            let departureTime = months[dtd.getMonth()] + ' ' + dtd.getDate() + ' at '+ controller.parseTime(dtd);


            let luggageText = '';
            switch (drive.optLuggageSize) {
                case 1:
                    luggageText = 'Small sized luggage only';
                    break;
                case 2:
                    luggageText = 'Medium sized luggage only';
                    break;
                case 3:
                    luggageText = 'Big luggage ok';
                    break;
            }

            const winterTiresText = drive.optWinterTires ? 'Has winter tires' : 'Does not have winter tires';
            const bicycleText = drive.optBicycle ? 'Bicycle ok' : 'No bicycle';
            const petsText = drive.optPets ? 'Pets ok' : 'No pets';

            // Set content
            document.getElementById('driver-first-name').textContent = driver.firstName;
            document.getElementById('driver-gender').textContent = driver.gender === 0 ? 'Male' : 'Female';
            document.getElementById('driver-age').textContent = driverAge;
            document.getElementById('drive-name').textContent = drive.start + ' to ' + drive.stop;
            document.getElementById('driver-driven').textContent = noDrives;
            document.getElementById('driver-rating').textContent = driverRating;
            document.getElementById('driver-reviews').textContent = driverReviews;
            document.getElementById('driver-driven').textContent = noDrives;
            document.getElementById('driver-rating').textContent = driverRating;
            document.getElementById('driver-reviews').textContent = driverReviews;
            document.getElementById('drive-name').textContent = drive.start + ' to ' + drive.stop;
            document.getElementById('drive-departure-time').textContent = departureTime;
            document.getElementById('drive-pickup').textContent = model.driveUser.start;
            document.getElementById('drive-pickup-time').textContent = model.driveUser.startTime;
            document.getElementById('drive-dropoff').textContent = model.driveUser.stop;
            document.getElementById('drive-comment').textContent = drive.comment.length > 0 ? '"' + drive.comment + '"' : '';
            document.getElementById('car-brand').textContent = drive.carBrand;
            document.getElementById('car-model').textContent = drive.carModel;
            document.getElementById('car-color').textContent = drive.carColor;
            document.getElementById('car-license-plate').textContent = drive.carLicensePlate;
            document.getElementById('car-license-plate-link').setAttribute('href', 'https://biluppgifter.se/fordon/' + drive.carLicensePlate);
            document.getElementById('opt-luggage-icon').classList.add(drive.optLuggageSize > 0 ? 'text-info' : 'text-muted');
            document.getElementById('opt-winter-tires-icon').classList.add(drive.optWinterTires ? 'text-info' : 'text-muted');
            document.getElementById('opt-bicycle-icon').classList.add(drive.optBicycle ? 'text-info' : 'text-muted');
            document.getElementById('opt-pets-icon').classList.add(drive.optPets ? 'text-info' : 'text-muted');
            document.getElementById('opt-luggage-text').textContent = luggageText;
            document.getElementById('opt-winter-tires-text').textContent = winterTiresText;
            document.getElementById('opt-bicycle-text').textContent = bicycleText;
            document.getElementById('opt-pets-text').textContent = petsText;
            document.getElementById('drive-passengers').innerHTML = passengerHtml;
            document.getElementById('drive-seats-left').textContent = drive.carNumberOfSeats - acceptedPassengers;
            document.getElementById('drive-content').classList.remove('d-none');
        }
    };

    const controller = {
        parseTime: date => {
            let hours = date.getHours();
            let minutes = date.getMinutes();
            hours = hours < 10 ? '0' + hours : hours;
            minutes = minutes < 10 ? '0' + minutes : minutes;

            return hours + ':' + minutes;
        },

        getDrive: () => window.base.rest.getDriveWrap(model.driveUser.driveId)
            .then(driveWrap => model.driveWrap = driveWrap),

        submitRequest: () => window.base.rest.requestSeat(model.driveUser.driveId, model.driveUser)
            .then(e => {
                if (e.error) {
                    view.showFailure(e.message, document.getElementById('request-alert-box'));
                } else {
                    view.showSuccess('We\'ve sent a request to the driver letting the person know you want a seat.', document.getElementById('request-alert-box'));
                    const requestButton = document.getElementById('drive-request');
                    requestButton.setAttribute('disabled', '');
                    requestButton.textContent = 'Requested';
                }
            }),

        submitReport: e => {
            e.preventDefault();

            model.driveReport.reportId = 0;
            model.driveReport.driveId = model.driveUser.driveId;
            model.driveReport.reportedByUser = model.user.userId;
            model.driveReport.reportMessage = document.getElementById('report-message').value;

            window.base.rest.reportDrive(model.driveUser.driveId, model.driveReport)
                .then(e => {
                    if (e.error) {
                        view.showFailure(e.message, document.getElementById('report-alert-box'));
                    } else {
                        view.showSuccess('Your report has been sent', document.getElementById('report-alert-box'));
                        document.getElementById('report-message').value = '';
                    }
                });
        },

        getUser: () => window.base.rest.getUser().then(u => {
            model.user = u;
            return model.driveUser.userId = u.userId;
        }),

        getDriverAndPassengers: () => {
            const driveUsers = model.driveWrap.users;

            return Promise.all(driveUsers.map((driveUser, i) => {
                return window.base.rest.getUser(driveUser.userId)
                    .then(user => model.driveWrap.users[i].info = user)
                    .then(() => {
                        if (driveUser.driver) {
                            return window.base.rest.getNumberOfDrivesForUser(driveUser.userId);
                        }

                        return 'N/A';
                    })
                    .then(noDrives => {
                        if (noDrives !== 'N/A') {
                            model.driveWrap.users[i].noDrives = noDrives;
                        }
                    })
            }));
        },

        loadQuery: searchQuery => {
            if (typeof searchQuery === 'undefined' && !Object.keys(model.driveUser).length) {
                window.base.changeLocation('/');
            }

            // Change the hash without firing hashchange
            history.pushState({}, '', '#/drive');

            document.getElementById('drive-request').onclick = controller.submitRequest;
            document.getElementById('report-modal-trigger').onclick = view.showReportModal;
            document.getElementById('report-form').onsubmit = controller.submitReport;

            model.driveUser = {
                driveId: searchQuery.driveId,
                start: searchQuery.tripStart,
                stop: searchQuery.tripStop,
                startTime: searchQuery.tripStartTime,
                driver: false,
                accepted: false,
                rated: false
            };

            controller.getUser()
                .then(() => controller.getDrive())
                .then(() => controller.getDriverAndPassengers())
                .then((() => view.renderDrive()))
        },

        load: () => controller.loadQuery()
    };

    return controller;
});